# Binary Classification 

Logistic Regression is used as a model to predict wheather a telco customer will churn. Below is the code used to analyse the data and to make customer classification.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# **Data Cleaning**

In [2]:
df = pd.read_csv("Telco-Customer-Churn.csv")

In [3]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [5]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"], errors="coerce")
df[df["TotalCharges"].isnull()]

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
488            No  No phone service             DSL                  Yes  ...   
753           Yes                No              No  No internet service  ...   
936           Yes                No             DSL                  Yes  ...   
1082          Yes               Yes              No  No internet service  ...   
1340           No  No phone service             DSL                  Yes  ...   
3331          Yes                No              No  No internet service  ...   
3826          Yes               Yes              No  No internet service  ...   
4380          Yes                No              No  No internet service  ...   
5218          Yes                No              No  No internet service  ...   
6670          Yes               Yes             DSL                   No  ...   
6754          Yes               Yes             DSL                  Yes  ...   

         DeviceProtection          TechSupport          StreamingTV  \
488                   Yes                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                   No                  Yes   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                   No                  Yes                   No   

          StreamingMovies  Contract PaperlessBilling  \
488                    No  Two year              Yes   
753   No internet service  Two year               No   
936                   Yes  Two year               No   
1082  No internet service  Two year               No   
1340                   No  Two year               No   
3331  No internet service  Two year               No   
3826  No internet service  Two year               No   
4380  No internet service  Two year               No   
5218  No internet service  One year              Yes   
6670                   No  Two year               No   
6754                   No  Two year              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges  Churn  
488   Bank transfer (automatic)          52.55           NaN     No  
753                Mailed check          20.25           NaN     No  
936                Mailed check          80.85           NaN     No  
1082               Mailed check          25.75           NaN     No  
1340    Credit card (automatic)          56.05           NaN     No  
3331               Mailed check          19.85           NaN     No  
3826               Mailed check          25.35           NaN     No  
4380               Mailed check          20.00           NaN     No  
5218               Mailed check     

In [6]:
df["TotalCharges"] = df["TotalCharges"].fillna(0)

In [7]:
df.columns = df.columns.str.lower().str.replace(" ", "_")

string_col = list(df.dtypes[df.dtypes == "object"].index)
for col in string_col:
    df[col] = df[col].str.lower().str.replace(" ", "_")
    
df.columns = df.columns.str.lower().str.replace(" ", "_")
df.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-vhveg  female              0     yes         no       1           no   
1  5575-gnvde    male              0      no         no      34          yes   
2  3668-qpybk    male              0      no         no       2          yes   
3  7795-cfocw    male              0      no         no      45           no   
4  9237-hqitu  female              0      no         no       2          yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  no_phone_service             dsl             no  ...               no   
1                no             dsl            yes  ...              yes   
2                no             dsl            yes  ...               no   
3  no_phone_service             dsl            yes  ...              yes   
4                no     fiber_optic             no  ...               no   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          no          no              no  month-to-month              yes   
1          no          no              no        one_year               no   
2          no          no              no  month-to-month              yes   
3         yes          no              no        one_year               no   
4          no          no              no  month-to-month              yes   

               paymentmethod monthlycharges  totalcharges  churn  
0           electronic_check          29.85         29.85     no  
1               mailed_check          56.95       1889.50     no  
2               mailed_check          53.85        108.15    yes  
3  bank_transfer_(automatic)          42.30       1840.75     no  
4           electronic_check          70.70        151.65    yes  

[5 rows x 21 columns]

In [8]:
df.churn = (df.churn == "yes").astype(int)

In [9]:
df.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  7590-vhveg  female              0     yes         no       1           no   
1  5575-gnvde    male              0      no         no      34          yes   
2  3668-qpybk    male              0      no         no       2          yes   
3  7795-cfocw    male              0      no         no      45           no   
4  9237-hqitu  female              0      no         no       2          yes   

      multiplelines internetservice onlinesecurity  ... deviceprotection  \
0  no_phone_service             dsl             no  ...               no   
1                no             dsl            yes  ...              yes   
2                no             dsl            yes  ...               no   
3  no_phone_service             dsl            yes  ...              yes   
4                no     fiber_optic             no  ...               no   

  techsupport streamingtv streamingmovies        contract paperlessbilling  \
0          no          no              no  month-to-month              yes   
1          no          no              no        one_year               no   
2          no          no              no  month-to-month              yes   
3         yes          no              no        one_year               no   
4          no          no              no  month-to-month              yes   

               paymentmethod monthlycharges  totalcharges  churn  
0           electronic_check          29.85         29.85      0  
1               mailed_check          56.95       1889.50      0  
2               mailed_check          53.85        108.15      1  
3  bank_transfer_(automatic)          42.30       1840.75      0  
4           electronic_check          70.70        151.65      1  

[5 rows x 21 columns]

## Setting up a validation framework

In [10]:
from sklearn.model_selection import train_test_split

df_full_train, df_test = train_test_split(df, test_size=0.2, random_state=42)
df_train, df_val = train_test_split(df_full_train, test_size=0.25, random_state=42)
len(df_train), len(df_val), len(df_test)

(4225, 1409, 1409)

In [11]:
df_train = df_train.reset_index(drop= True)
df_val = df_val.reset_index(drop= True)
df_test = df_test.reset_index(drop= True)

y_train = df_train.churn.values
y_val = df_val.churn.values
y_test = df_test.churn.values

del df_train["churn"]
del df_val["churn"]
del df_test["churn"]

In [12]:
df_train.head()

customerid  gender  seniorcitizen partner dependents  tenure phoneservice  \
0  6776-tlwoi    male              0      no         no       3          yes   
1  0036-ihmot  female              0     yes        yes      55          yes   
2  2843-cqmeg    male              0      no         no      24          yes   
3  3247-mhjkm    male              0      no         no       1          yes   
4  1194-bhjyc    male              0     yes         no      61           no   

      multiplelines internetservice       onlinesecurity         onlinebackup  \
0                no              no  no_internet_service  no_internet_service   
1                no     fiber_optic                   no                  yes   
2                no             dsl                   no                  yes   
3                no              no  no_internet_service  no_internet_service   
4  no_phone_service             dsl                  yes                   no   

      deviceprotection          techsupport          streamingtv  \
0  no_internet_service  no_internet_service  no_internet_service   
1                  yes                  yes                  yes   
2                   no                   no                   no   
3  no_internet_service  no_internet_service  no_internet_service   
4                  yes                  yes                  yes   

       streamingmovies        contract paperlessbilling  \
0  no_internet_service  month-to-month               no   
1                  yes        one_year              yes   
2                   no  month-to-month              yes   
3  no_internet_service  month-to-month               no   
4                  yes        two_year              yes   

               paymentmethod  monthlycharges  totalcharges  
0               mailed_check           19.85         64.55  
1  bank_transfer_(automatic)          103.70       5656.75  
2               mailed_check           49.70       1218.25  
3               mailed_check           20.20         20.20  
4               mailed_check           62.15       3778.85

## EDA

In [13]:
df_full_train = df_full_train.reset_index(drop=True)
df_full_train.isnull().sum()

customerid          0
gender              0
seniorcitizen       0
partner             0
dependents          0
tenure              0
phoneservice        0
multiplelines       0
internetservice     0
onlinesecurity      0
onlinebackup        0
deviceprotection    0
techsupport         0
streamingtv         0
streamingmovies     0
contract            0
paperlessbilling    0
paymentmethod       0
monthlycharges      0
totalcharges        0
churn               0
dtype: int64

In [14]:
global_churn = df_full_train["churn"].value_counts(normalize=True).iloc[1]
global_churn

0.2655307064252751

Churn rate is 26%

In [15]:
numeric = ["tenure", "monthlycharges", 'totalcharges']
categorical = ['gender', 'seniorcitizen', 'partner', 'dependents', 'phoneservice', 'multiplelines', 'internetservice',
       'onlinesecurity', 'onlinebackup', 'deviceprotection', 'techsupport',
       'streamingtv', 'streamingmovies', 'contract', 'paperlessbilling',
       'paymentmethod']

## Feature Importance

### Churn Rate

In [16]:
from IPython.display import display

for col in categorical:
    print(col)
    df_group = df_full_train.groupby(col)["churn"].agg(["mean", "count"])
    df_group["diff"] = global_churn - df_group["mean"] 
    df_group["risk"] = df_group["mean"] / global_churn
    display(df_group)

gender


mean  count      diff      risk
gender                                     
female  0.270841   2747 -0.005310  1.019998
male    0.260478   2887  0.005053  0.980971

seniorcitizen


mean  count      diff      risk
seniorcitizen                                     
0              0.237098   4728  0.028433  0.892922
1              0.413907    906 -0.148377  1.558793

partner


mean  count      diff      risk
partner                                     
no       0.326446   2904 -0.060916  1.229411
yes      0.200733   2730  0.064798  0.755968

dependents


mean  count      diff      risk
dependents                                     
no          0.312326   3951 -0.046795  1.176233
yes         0.155674   1683  0.109856  0.586276

phoneservice


mean  count      diff      risk
phoneservice                                     
no            0.253623    552  0.011908  0.955156
yes           0.266824   5082 -0.001293  1.004871

multiplelines


mean  count      diff      risk
multiplelines                                        
no                0.251397   2685  0.014134  0.946771
no_phone_service  0.253623    552  0.011908  0.955156
yes               0.284105   2397 -0.018574  1.069952

internetservice


mean  count      diff      risk
internetservice                                     
dsl              0.191851   1939  0.073679  0.722521
fiber_optic      0.415558   2481 -0.150028  1.565010
no               0.076606   1214  0.188924  0.288502

onlinesecurity


mean  count      diff      risk
onlinesecurity                                          
no                   0.416014   2810 -0.150484  1.566727
no_internet_service  0.076606   1214  0.188924  0.288502
yes                  0.145342   1610  0.120189  0.547363

onlinebackup


mean  count      diff      risk
onlinebackup                                            
no                   0.398693   2448 -0.133162  1.501494
no_internet_service  0.076606   1214  0.188924  0.288502
yes                  0.216531   1972  0.048999  0.815467

deviceprotection


mean  count      diff      risk
deviceprotection                                        
no                   0.387706   2489 -0.122175  1.460117
no_internet_service  0.076606   1214  0.188924  0.288502
yes                  0.226825   1931  0.038705  0.854234

techsupport


mean  count      diff      risk
techsupport                                             
no                   0.413472   2791 -0.147941  1.557153
no_internet_service  0.076606   1214  0.188924  0.288502
yes                  0.152855   1629  0.112676  0.575657

streamingtv


mean  count      diff      risk
streamingtv                                             
no                   0.335418   2239 -0.069887  1.263197
no_internet_service  0.076606   1214  0.188924  0.288502
yes                  0.298945   2181 -0.033415  1.125841

streamingmovies


mean  count      diff      risk
streamingmovies                                         
no                   0.333333   2223 -0.067803  1.255348
no_internet_service  0.076606   1214  0.188924  0.288502
yes                  0.301320   2197 -0.035789  1.134784

contract


mean  count      diff      risk
contract                                           
month-to-month  0.426533   3083 -0.161002  1.606340
one_year        0.117987   1212  0.147544  0.444343
two_year        0.028379   1339  0.237151  0.106878

paperlessbilling


mean  count      diff      risk
paperlessbilling                                    
no                0.16414   2309  0.101390  0.618159
yes               0.33594   3325 -0.070409  1.265164

paymentmethod


mean  count      diff      risk
paymentmethod                                                 
bank_transfer_(automatic)  0.174475   1238  0.091056  0.657080
credit_card_(automatic)    0.152404   1227  0.113126  0.573961
electronic_check           0.449921   1887 -0.184390  1.694420
mailed_check               0.190328   1282  0.075203  0.716782

### Mutual Information 
concept from information theory, it tells us how much we can learn about one variable if we know the value from another

In [17]:
from sklearn.metrics import mutual_info_score

def mutual_info_scores(series):
    return mutual_info_score(series, df_full_train.churn)

mutual_info = df_full_train[categorical].apply(mutual_info_scores)
mutual_info.sort_values(ascending=False)

contract            0.096652
onlinesecurity      0.063393
techsupport         0.060935
internetservice     0.053313
onlinebackup        0.045424
paymentmethod       0.042861
deviceprotection    0.042007
streamingtv         0.030844
streamingmovies     0.030705
paperlessbilling    0.019077
dependents          0.014129
partner             0.010227
seniorcitizen       0.010059
multiplelines       0.000654
gender              0.000069
phoneservice        0.000040
dtype: float64

In [18]:
mutual_info.sort_values(ascending=False).index

Index(['contract', 'onlinesecurity', 'techsupport', 'internetservice',
       'onlinebackup', 'paymentmethod', 'deviceprotection', 'streamingtv',
       'streamingmovies', 'paperlessbilling', 'dependents', 'partner',
       'seniorcitizen', 'multiplelines', 'gender', 'phoneservice'],
      dtype='object')

### Correlation

In [19]:
df_full_train[numeric].corrwith(df_full_train.churn).abs().sort_values(ascending=False)

tenure            0.344925
totalcharges      0.193370
monthlycharges    0.188574
dtype: float64

In [20]:
from sklearn.feature_extraction import DictVectorizer

selected_categorical = ['contract', 'onlinesecurity', 'techsupport', 'internetservice', 'onlinebackup', 'paymentmethod', 'deviceprotection', 'streamingtv','streamingmovies']
selected_numeric = ["tenure", "totalcharges",]

def transform_cols(dataframe):
    dv = DictVectorizer(sparse=False)
    dic = dataframe[selected_categorical+selected_numeric].to_dict(orient="records")
    dv.fit(dic)
    return dv.transform(dic)

X_train = transform_cols(df_train)
X_val = transform_cols(df_val)
X_test = transform_cols(df_test)

## Model Training

In [21]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(solver='liblinear', random_state=1)
model.fit(X_train, y_train)

val_preds = model.predict(X_val)
(val_preds == y_val).mean()

0.7920511000709723

In [22]:
val_preds_2 = model.predict_proba(X_val)[:, 1]
churn_decisions = val_preds_2 >= 0.5
(churn_decisions.astype(int) == y_val).mean()

0.7920511000709723

In [23]:
train_preds = model.predict(X_train)
(train_preds == y_train).mean()

0.8004733727810651

## Use the model

In [24]:
X_full = transform_cols(df_full_train)
y_full = df_full_train.churn.values

In [25]:
lr = LogisticRegression(solver='liblinear', random_state=1)
lr.fit(X_full, y_full)

test_preds = lr.predict(X_test)
(test_preds == y_test).mean()

0.8147622427253371

Accurancy of the model tested on the test dataset is 81%